Kk/scsb update metadata

emu47 opened the pull request:
Ok so this has been baking for too long in the dark obviously. Time to bring it into the light. 

Willing to step through it tomorrow with anyone who is interested. Refile methods / actions / views are still in progress and will be fully finished in a future branch. 

Meat of this pull request is to look over and find issues with the update and the transfer actions / views on the item controller.

Also adds some first tests to our app.

emu47 committed:
[KK] Incorporating design toolkit gem.

emu47 committed:
[KK] Configuration settings for sqs dead letter office.

emu47 committed:
[KK] Updating ruby version and adding rspec gem.

emu47 committed:
[KK] Recreating the legacy react site fucntionality in rails.

emu47 committed:
[KK] Adding rspec hidden file.

emu47 committed:
[KK] Resolving conflicts with master.

emu47 committed:
[KK] Adding validations for bib record numbers and fixing configuration for docker.

emu47 committed:
[KK] Fixing conflicts with master.

emu47 committed:
[KK] removing dotenv from gemfile.

nodanaonlyzuul committed:
Have travis run tests

nodanaonlyzuul commented on app/controllers/items_controller.rb:
```diff
@@ -2,18 +2,86 @@ class ItemsController < OauthController
+      invalid_barcodes = []
+    elsif message.valid_barcodes.present? 
+      invalid_barcodes = message.invalid_barcodes
+      message.barcodes = message.valid_barcodes
```
We talked about `Message#send_update_message_to_sqs` owning the logic of only sending valid/well-formed barcodes through instead of the whole `barcodes` Array.

nodanaonlyzuul commented on app/controllers/items_controller.rb:
```diff
@@ -2,18 +2,86 @@ class ItemsController < OauthController
   end
 
   def send_metadata
-    barcodes = params[:barcodes]
```
We talked about not needing an `invalid_barcodes` local variable because we can rely on `Message#invalid_barcodes`. We may want to add a few more specs to the `Message` unit tests about how `invalid_barcodes`'s state starts and changes.

nodanaonlyzuul commented on app/controllers/items_controller.rb:
```diff
@@ -2,18 +2,86 @@ class ItemsController < OauthController
+    if message.valid_barcodes.present? && invalid_barcodes.blank?
+      flash[:success] = "The barcode(s) have been submitted for processing."
+    end
+    redirect_to action: 'update_metadata', notice: { barcodes: invalid_barcodes, protect_cgd: protect_cgd}
```
I think we can find a way to smuggle these values back to the "edit" form without using the flash.
Let's look into a way of using an `ActiveModel` object and either `form_for` or the `value: ` option in the form tags to have the form get its values directly from our instance of `message.`

nodanaonlyzuul replied:
Ahhh - this isn't actually using the flash - it's turning these into parameters. 
For example, after a bad submit you are redirected to:

```
http://local.nypl.org:3001/update_metadata?notice[barcodes][]=sdahusdhiu¬ice[protect_cgd]=
```

nodanaonlyzuul committed:
Moving 'require' statement to top of file

nodanaonlyzuul committed:
Moving require statement to top of file

nodanaonlyzuul committed:
Remove home-rolled contructors and using ActiveModel::Model

```
ActiveModel::Model "Includes the required interface for an object
to interact with Action Pack and Action View, using different Active Model modules.
It includes model name introspections, conversions, translations and validations.
Besides that, it allows you to initialize the object with a hash of attributes,
pretty much like Active Record does.
```

source: https://api.rubyonrails.org/classes/ActiveModel/Model.html

nodanaonlyzuul commented on Gemfile:
```diff
@@ -23,7 +23,7 @@ gem 'coffee-rails', '~> 4.2'
 # Use ActiveModel has_secure_password
 # gem 'bcrypt', '~> 3.1.7'
-gem 'aws-sdk'
+gem 'aws-sdk-sqs', '~> 1.3'
```
THANK YOU!

nodanaonlyzuul committed:
Don't keep simplecov output in source control

nodanaonlyzuul commented on app/models/message.rb:
```diff
@@ -0,0 +1,102 @@
+    entries = [
+        {
+          id: self.barcodes.first,
+          message_body: "{
```
For `send_update_message_to_sqs` AND `send_transfer_message_to_sqs`.

Two things:

1.  If we are already storing stuff from the OAuth response in the session, you can probably get the current user's email for this.

2.  If you want to get out of hand-rolling JSON....MAYYYBEEEE this will take a hash and JSONify it OR.....
call `JSON.generate({action: self.action, barcodes: self.barcodes})`

nodanaonlyzuul committed:
Remove inline require of 'aws-sdk-sqs'

emu47 committed:
[KK] no more invalid_barcodes instance variable. also cleaned up gemfile.

emu47 committed:
[KK] Adjusting the parameters to avoid stepping on each other.

emu47 committed:
[KK] Fixing protect_cgd attribute handling.

emu47 committed:
[KK] Fixing protect cgd for transfer action

emu47 committed:
Merge branch 'master' into kk/scsb-update-metadata

emu47 committed:
[KK] Removing temp email

emu47 committed:
[KK] Factoring sqs connections into a library.

nodanaonlyzuul committed:
Remove commented out code

emu47 committed:
[KK] Using new endpoint Ho-Ling made and reducing redundancy in env variables per Kang suggestion.

emu47 committed:
[KK] Fixing conflict.

nodanaonlyzuul commented on .env.example:
```diff
@@ -1,20 +1,27 @@
+API_BASE_URL=The base URL of the APIs of NYPL's microservices on development platform
```
@emu47 - thank you for putting this in alphabetical order.

![tenor](https://user-images.githubusercontent.com/8731/44227625-0efa8980-a161-11e8-8144-044aaed27ec1.gif)
